include/armpp/hal/nvic.hpp

```diff
@@ -0,0 +1,162 @@
+    static constexpr address icpr_base    = 0xe000e280;
+    static constexpr address iabr_base    = 0xe000e300;
+    static constexpr address ip_base      = 0xe000e400;
+    static constexpr address stir_base    = 0xe000ef00;
+
+    interrupt_set_enable_register       iser_;    // 0xe000e100 - 0xe000e120
+    raw_register const                  reserved_0_[24];
+    interrupt_clear_enable_register     icer_;    // 0xe000e180 - 0e000e1a0
+    raw_register const                  reserved_1_[24];
+    interrupt_set_pending_register      ispr_;    // 0xe000e200 - 0xe000e220
+    raw_register const                  reserved_2_[24];
+    interrupt_clear_pending_register    icpr_;    // 0xe000e280 - 0xe000e2a0
+    raw_register const                  reserved_3_[24];
+    active_bit_register                 iabr_;    // 0xe000e300 - 0xe000e320
+    raw_register const                  reserved_4_[56];
+    interrupt_priority_register         ip_;    // 0xe000e400 - 0xe000e41f (?)
+    raw_register const                  reserved_5_[644];
+    software_trigger_interrupt_register stir_;
+};
+static_assert(offsetof(nvic_registers, iser_)
+              == (nvic_registers::iser_base - nvic_registers::base_address));
+static_assert(offsetof(nvic_registers, icer_)
+              == (nvic_registers::icer_base - nvic_registers::base_address));
+static_assert(offsetof(nvic_registers, ispr_)
+              == (nvic_registers::ispr_base - nvic_registers::base_address));
+static_assert(offsetof(nvic_registers, icpr_)
+              == (nvic_registers::icpr_base - nvic_registers::base_address));
+static_assert(offsetof(nvic_registers, iabr_)
+              == (nvic_registers::iabr_base - nvic_registers::base_address));
+static_assert(offsetof(nvic_registers, ip_)
+              == (nvic_registers::ip_base - nvic_registers::base_address));
+static_assert(offsetof(nvic_registers, stir_)
+              == (nvic_registers::stir_base - nvic_registers::base_address));
+
+class nvic : nvic_registers {
+public:
+    using nvic_registers::base_address;
+
+public:
+    nvic()            = delete;
+    nvic(nvic const&) = delete;
+    nvic(nvic&&)      = delete;
+
+    nvic&
+    operator=(nvic const&)
+        = delete;
+    nvic&
+    operator=(nvic&&)
+        = delete;
+
+public:
+    void
+    enable_irq(std::size_t index)
+    {
+        iser_.set[index] = set_t::set;
+    }
+
+    void
+    disable_irq(std::size_t index)
+    {
+        icer_.set[index] = reset_t::reset;
+    }
+
+    bool
+    irq_enabled(std::size_t index) const
+    {
+        return iser_.get[index] == enabled_t::enabled;
+    }
+
+    void
+    set_pending(std::size_t index)
+    {
+        ispr_.set[index] = set_t::set;
+    }
+
+    void
+    clear_pending(std::size_t index)
+    {
+        icpr_.set[index] = reset_t::reset;
+    }
+
+    bool
+    is_pending(std::size_t index) const
+    {
+        return ispr_.get[index] == active_t::active;
+    }
```

Comment:
**Suggestion:** Consider adding checks for the index in the methods of the `nvic` class. This will prevent potential out-of-bounds access if an invalid index is passed.
```suggestion
    void
    enable_irq(std::size_t index)
    {
        if (index < interrupt_count) {
            iser_.set[index] = set_t::set;
        }
    }

    void
    disable_irq(std::size_t index)
    {
        if (index < interrupt_count) {
            icer_.set[index] = reset_t::reset;
        }
    }

    bool
    irq_enabled(std::size_t index) const
    {
        if (index < interrupt_count) {
            return iser_.get[index] == enabled_t::enabled;
        }
        return false;
    }

    void
    set_pending(std::size_t index)
    {
        if (index < interrupt_count) {
            ispr_.set[index] = set_t::set;
        }
    }

    void
    clear_pending(std::size_t index)
    {
        if (index < interrupt_count) {
            icpr_.set[index] = reset_t::reset;
        }
    }

    bool
    is_pending(std::size_t index) const
    {
        if (index < interrupt_count) {
            return ispr_.get[index] == active_t::active;
        }
        return false;
    }
```